File: apple/HopBearers/Sources/HopBearerCore/BearerManager.swift
// BearerManager — the bearer registry. The consumer registers one or more `Bearer`s (BLE first; LAN /
// Wi-Fi Direct / relay / a third-party transport later) and then talks ONLY to the manager. The
// manager is ITSELF a `Bearer`, so the consumer drives the whole transport layer through the exact
// same interface it would use for a single bearer:
//
//     let mgr = BearerManager()
//     mgr.register(BleBearer(myId: myId))
//     mgr.register(LanBearer(...))        // any future Bearer registers identically
//     mgr.sink = myConsumer               // links from ALL bearers surface here
//     mgr.start()                         // fans out to every registered bearer
//
// Two jobs, both purely in terms of the `Bearer`/`LinkSink` contract — NO BLE/Wi-Fi/socket types ever
// cross this boundary:
//   1. Fan-out: start()/stop() reach every bearer; send(_:on:) routes to the bearer owning the link.
//   2. One link-id space: each bearer mints its own (local) LinkIds starting from 1, so two bearers
//      would otherwise collide. The manager mints a process-global LinkId per link and translates each
//      bearer's local id into that global space, so the consumer keys all its state on ONE id space
//      regardless of which radio a link rode in on.
//
// Anyone writing a new transport implements `Bearer` and registers it — nothing here knows or cares
// what it is. That is the whole point: the interface is the same no matter the bearer.

import Foundation

/// The registry + multiplexer. Register bearers, set `sink`, then drive everything as one `Bearer`.
///
/// Threading: the manager's routing maps are mutated from the lane callbacks (which arrive on each
/// bearer's work queue) and read in `send`. Today every bearer shares the one `bleRunLoop`/`.main`
/// serial context (CLI) or the production consumer's single core queue (Stage C), so these are already
/// serialized by the host's threading model — the same single-threaded invariant `BleBearer` itself
/// relies on. A future bearer on an independent queue would need this funneled through a serial queue.
public final class BearerManager: Bearer {
    /// The consumer's sink — every link, from every registered bearer, surfaces here in ONE global
    /// LinkId space. Weak: the consumer typically owns the manager, so a strong ref would cycle.
    public weak var sink: LinkSink?

    // THREADING: each bearer runs on its OWN queue (the BLE bearer's run loop, the LAN bearer's serial
    // queue, …) and the consumer pings from yet another, so the routing maps are touched concurrently.
    // A lock guards every map access; sink callbacks are always invoked OUTSIDE the lock (so a consumer
    // that calls back into `send` from linkUp/linkBytes can't deadlock).
    private let lock = NSLock()
    private var bearers: [Bearer] = []
    private var lanes: [Lane] = []                                    // keep per-bearer shims alive (bearer.sink is weak)
    private var nextGlobal: LinkId = 1                                // global link-id space (distinct from any bearer's)
    private var toGlobal: [ObjectIdentifier: [LinkId: LinkId]] = [:]  // bearer -> (localLink -> globalLink)
    private var fromGlobal: [LinkId: (Bearer, LinkId)] = [:]          // globalLink -> (owning bearer, localLink)

    public init() {}

    /// Register a bearer. The manager installs itself as the bearer's sink (via a private per-bearer
    /// lane that tags + translates the bearer's link ids), so from here on the bearer's links flow up
    /// through the manager. Call before `start()`.
    public func register(_ bearer: Bearer) {
        let lane = Lane(manager: self, bearer: bearer)
        bearer.sink = lane                                           // bearer holds this weakly...
        lock.lock(); lanes.append(lane); bearers.append(bearer); lock.unlock()  // ...manager keeps it alive
    }

    // MARK: Bearer — the uniform interface the consumer drives (fans out to every registered bearer)

    public func start() { snapshotBearers().forEach { $0.start() } }
    public func stop()  { snapshotBearers().forEach { $0.stop() } }

    public func send(_ bytes: Data, on link: LinkId) {
        lock.lock(); let route = fromGlobal[link]; lock.unlock()
        guard let (bearer, local) = route else { return }            // unknown / already-closed global link
        bearer.send(bytes, on: local)
    }

    private func snapshotBearers() -> [Bearer] { lock.lock(); defer { lock.unlock() }; return bearers }

    // MARK: lane callbacks — bearer-local link ids in, global link ids out (map work locked, sink unlocked)

    fileprivate func up(_ bearer: Bearer, _ local: LinkId, _ role: LinkRole, _ peerId: Data) {
        lock.lock()
        let g = nextGlobal; nextGlobal += 1
        toGlobal[ObjectIdentifier(bearer), default: [:]][local] = g
        fromGlobal[g] = (bearer, local)
        lock.unlock()
        sink?.linkUp(g, role: role, peerId: peerId)
    }

    fileprivate func bytes(_ bearer: Bearer, _ local: LinkId, _ data: Data) {
        lock.lock(); let g = toGlobal[ObjectIdentifier(bearer)]?[local]; lock.unlock()
        guard let g else { return }
        sink?.linkBytes(g, data)
    }

    fileprivate func down(_ bearer: Bearer, _ local: LinkId) {
        let oid = ObjectIdentifier(bearer)
        lock.lock()
        let g = toGlobal[oid]?[local]
        if let g { toGlobal[oid]?[local] = nil; fromGlobal[g] = nil }
        lock.unlock()
        guard let g else { return }
        sink?.linkDown(g)
    }
}

/// One per registered bearer: the `LinkSink` a bearer reports to. It tags each callback with its owning
/// bearer and forwards to the manager, which does the local→global LinkId translation. Holds the
/// manager + bearer `unowned` — the manager owns both this lane and the bearer, so neither outlives it.
private final class Lane: LinkSink {
    unowned let manager: BearerManager
    unowned let bearer: Bearer
    init(manager: BearerManager, bearer: Bearer) { self.manager = manager; self.bearer = bearer }
    func linkUp(_ link: LinkId, role: LinkRole, peerId: Data) { manager.up(bearer, link, role, peerId) }
    func linkBytes(_ link: LinkId, _ bytes: Data) { manager.bytes(bearer, link, bytes) }
    func linkDown(_ link: LinkId) { manager.down(bearer, link) }
}

File: apps/ble-lab/android/bearer-core/src/main/java/sh/hopme/bearers/BearerManager.kt
package sh.hopme.bearers

// BearerManager, the bearer registry, the Android mirror of apple/HopBearers' BearerManager.swift.
// The consumer registers one or more Bearers (BLE first; LAN / Wi-Fi Direct / relay later) and then
// talks ONLY to the manager. The manager is ITSELF a Bearer, so the consumer drives the whole
// transport layer through the same interface it would use for a single bearer:
//
//     val mgr = BearerManager()
//     mgr.register(BleBearer(ctx, myId))
//     mgr.sink = myConsumer        // links from ALL bearers surface here
//     mgr.start()                  // fans out to every registered bearer
//
// Two jobs, purely in terms of Bearer/LinkSink, NO BLE/Wi-Fi/socket types ever cross this boundary:
//   1. Fan-out: start()/stop() reach every bearer; send() routes to the bearer owning the link.
//   2. One link-id space: each bearer mints its own (local) LinkIds from 1, so two bearers would
//      collide. The manager mints a process-global LinkId per link and translates each bearer's local
//      id into that global space, so the consumer keys all state on ONE id space regardless of radio.
//
// THREADING (the key Android difference from Apple): Apple's manager runs on a single run loop, so its
// maps need no locks. Here the lane callbacks arrive from MULTIPLE bearer threads (BleBearer surfaces
// linkUp/linkBytes/linkDown from per-link l2cap-rx threads, the main Handler, and the accept thread).
// So every routing-map mutation/read is guarded by a single `lock`.

import java.util.concurrent.CopyOnWriteArrayList

/// The registry + multiplexer. Register bearers, set `sink`, then drive everything as one Bearer.
// `baseLinkId` offsets this manager's global link-id space. A host that mints link ids for OTHER
// transports outside the manager (the production node's relay/Multipeer links share one nextLinkId
// counter) passes a high base so manager ids can never collide. Default 1 (standalone / clean-room).
class BearerManager(baseLinkId: LinkId = 1) : Bearer {
    /// The consumer's sink: every link, from every registered bearer, surfaces here in ONE global
    /// LinkId space.
    override var sink: LinkSink? = null

    /// Bearer contract. The manager aggregates many transports, so it has no single tag of its own; this
    /// is only read if a manager were nested inside another (it isn't). Per-link tags come from
    /// `transportNameOf(link)`, which reports the OWNING bearer's tag.
    override val transportName = "Mesh"

    private val lock = Any()
    private val bearers = CopyOnWriteArrayList<Bearer>() // iterated lock-free by start()/stop()
    private val lanes = CopyOnWriteArrayList<Lane>()     // keep per-bearer shims alive
    private var nextGlobal: LinkId = baseLinkId          // global link-id space (distinct from any bearer's)
    private val fromGlobal = HashMap<LinkId, Pair<Bearer, LinkId>>() // global -> (owning bearer, local)

    /// Register a bearer. The manager installs itself as the bearer's sink (via a private per-bearer
    /// lane that tags + translates the bearer's link ids). Call before `start()`.
    fun register(bearer: Bearer) {
        val lane = Lane(bearer)
        bearer.sink = lane
        lanes.add(lane)
        bearers.add(bearer)
    }

    // ---- Bearer: the uniform interface the consumer drives (fans out to every registered bearer) ----

    override fun start() = bearers.forEach { it.start() }
    override fun stop() = bearers.forEach { it.stop() }

    override fun send(bytes: ByteArray, link: LinkId) {
        val target = synchronized(lock) { fromGlobal[link] } ?: return // unknown / already-closed global link
        target.first.send(bytes, target.second)
    }

    // ---- transport labelling, the consumer's UI tags each shared-bearer link by its REAL transport ----

    /// The short transport tag (e.g. "BT"/"LAN"/"Wi-Fi Direct"/"Relay") of the bearer owning `link`, or
    /// null if the link is unknown / already closed. Cosmetic only, used to label a shared-bearer peer.
    /// (Named `…Of` rather than `transportName(link)` to avoid colliding with the Bearer `transportName`
    /// property above; Kotlin can't disambiguate a same-named property and function at a call site.)
    fun transportNameOf(link: LinkId): String? =
        synchronized(lock) { fromGlobal[link] }?.first?.transportName

    /// Every currently-up global link grouped by its owning bearer's transport tag → count. The consumer
    /// builds its per-transport status list from this when shared bearers are on.
    fun activeTransports(): Map<String, Int> {
        val routes = synchronized(lock) { fromGlobal.values.toList() }
        val out = HashMap<String, Int>()
        for ((bearer, _) in routes) out[bearer.transportName] = (out[bearer.transportName] ?: 0) + 1
        return out
    }

    // ---- lane callbacks, bearer-local link ids in, global link ids out ----

    private fun up(lane: Lane, local: LinkId, role: LinkRole, peerId: ByteArray) {
        val g: LinkId
        synchronized(lock) {
            g = nextGlobal; nextGlobal += 1
            lane.toGlobal[local] = g
            fromGlobal[g] = Pair(lane.bearer, local)
        }
        sink?.linkUp(g, role, peerId)
    }

    private fun bytes(lane: Lane, local: LinkId, data: ByteArray) {
        val g = synchronized(lock) { lane.toGlobal[local] } ?: return
        sink?.linkBytes(g, data)
    }

    private fun down(lane: Lane, local: LinkId) {
        var g: LinkId? = null
        synchronized(lock) {
            lane.toGlobal[local]?.let { gg ->
                lane.toGlobal.remove(local)
                fromGlobal.remove(gg)
                g = gg
            }
        }
        g?.let { sink?.linkDown(it) }
    }

    /// One per registered bearer: the LinkSink a bearer reports to. It tags each callback with its
    /// owning bearer (and owns that bearer's local->global map) and forwards to the manager, which does
    /// the translation. Mirrors Apple's `Lane`.
    private inner class Lane(val bearer: Bearer) : LinkSink {
        val toGlobal = HashMap<LinkId, LinkId>() // localLink -> globalLink (guarded by manager.lock)
        override fun linkUp(link: LinkId, role: LinkRole, peerId: ByteArray) = up(this, link, role, peerId)
        override fun linkBytes(link: LinkId, bytes: ByteArray) = bytes(this, link, bytes)
        override fun linkDown(link: LinkId) = down(this, link)
    }
}
